homeassistant/components/mqtt/discovery.py

```diff
@@ -163,11 +163,9 @@ class MQTTConfig(dict):
                 and component in IMPLICIT_STATE_TOPIC_COMPONENTS
             ):
                 # state_topic not specified, infer from discovery topic
-                payload[CONF_STATE_TOPIC] = "{}/{}/{}{}/state".format(
-                    discovery_topic,
-                    component,
-                    "%s/" % node_id if node_id else "",
-                    object_id,
+                payload[CONF_STATE_TOPIC] = (
+                    f"{discovery_topic}/{component}/"
+                    f"{'%s/' % node_id if node_id else ''}{object_id}/state"
```

Comment:
Don't mix in old style string formatting.

Reply:
Not sure what to do with '%s/' is this correct?
```suggestion
                    f"{node_id if node_id else ''}/{object_id}/state"
```

Reply:
```py
node_id or ''
```
Should work too.

Reply:
```suggestion
                    f"{node_id or ''}/{object_id}/state"
```